index.rst

```diff
@@ -13,17 +13,151 @@
+When schema for a manager is created or upgraded, its current version is stored in a special table with a name ``butler_attributes`` which also records other metadata in a key-value format.
+Compared to regular semantic versioning, versions in ``daf_butler`` have somewhat different compatibility rules:
+
+- versions that differ in ``MAJOR`` number are not compatible, just like in semantic versioning,
+- versions that differ in ``MINOR`` number are backward-compatible, but only for reading, and not compatible for write access,
+- versions that differ in ``PATCH`` number only are fully compatible.
+
+Schema upgrade tools for Butler database are implemented on top of `Alembic`_ which is based on ``SQLAlchemy``.
+As Butler schema is more complicated that a typical Alembic-managed database and has non-trivial migration paths, a special Alembic-based tool `daf_butler_migrate`_ was developed to support schema migrations.
+`DMTN-191`_ provides additional information about schema migration for Butler database.
+
+
+Schema versioning for APDB
+==========================
+
+To handle schema versioning for APDB it needs a mechanism to identify and record versions and the tools to check version compatibility and perform schema migrations for existing databases.
+
+As mentioned above, APDB schema is a product of two entities - ``sdm_schemas`` definitions and ``dax_apdb`` code - each of these two can evolve independently.
+To identify database schema version, the versions of both ``sdm_schemas`` and ``dax_apdb`` have to be known.
+Presently ``sdm_schemas`` (or rather `felis`_) does not provide a way to define or access version numbers for the schema that it defines, so it will have to be extended to allow specification of the version number, and possibly to include some form of compatibility rules.
+Similarly, ``dax_apdb`` does not provide a way to specify its code version, it has to be extended as well to include that information (independently for SQL and Cassandra implementations).
+
+
+Version recording
+-----------------
+
+The database needs to record the versions with which it was created (or later upgraded).
+One common approach for this is to define a separate metadata table that can record various additional information about the database itself.
+This table can be a simple key-value storage wit two string columns, indexed by a key value, this approach is used for ``daf_butler`` schema management tools.
+Both versions used to create database schema will be recorded in the metadata table, one possible example choosing names for keys could be::
+
+    key                           | value
+   -------------------------------+----------
+    version:sdm_schemas/apdb.yaml | 1.2.0
+    version:dax_apdb/ApdbSql      | 2.0.1
+
+Addition of this metadata table to the existing databases would be a schema change in itself, which could probably be managed by the migration tools.
```

Comment:
I understand there is a bunch of APDB schemas already at USDF, I wanted to avoid disruption for people running their own tests (maybe even on SQLite files) when new release suddenly appears that requires everyone do something.

Reply:
If it's easy enough to build said migration tool for version 0, that's fine. I don't think it'll be that disruptive: we already have a breaking schema change from the RFC-906 a few months ago, and ew are just using either an older or newer pipelines release to work with the data, depending on which schema it is. The SQLite files from tests get regenerated quite often, and are generally not considered important enough to maintain compatibility with.

The only one that's vaguely important is the test file in `analysis_ap`, but that one still needs updating for the RFC-906 schema change anyway (and the package is not yet in lsst_distrib). We could roll this metadata addition into the other updates that are necessary for that package.